### PR TITLE
EventGrid: CLI support for 2019-06-01 API Version

### PR DESCRIPTION
**Pull request by ahamad-MS**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by ahamad-MS** [comment]
@yugangw-msft   Can you please assign someone for this PR. thanks

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py [hunk 1]
```diff
@@ -32,12 +46,30 @@ def load_command_table(self, _):
...
+                                 setter_name='update',
+                                 client_factory=domains_factory)
+
     custom_tmpl = 'azure.cli.command_modules.eventgrid.custom#{}'
     eventgrid_custom = CliCommandType(operations_tmpl=custom_tmpl)
 
     with self.command_group('eventgrid event-subscription', client_factory=event_subscriptions_factory) as g:
         g.custom_command('create', 'cli_eventgrid_event_subscription_create')
-        g.custom_show_command('show', 'cli_eventgrid_event_subscription_get')
+        g.custom_command('show', 'cli_eventgrid_event_subscription_get')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Should use custom_show_command. Why was this changed?

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py [hunk 2]
```diff
@@ -49,5 +81,5 @@ def load_command_table(self, _):
 
     with self.command_group('eventgrid topic-type', topic_type_mgmt_util) as g:
         g.command('list', 'list')
-        g.show_command('show', 'get')
+        g.command('show', 'get')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Should be `show_command`. Why was this changed?

Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Fixed

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py (hunk 1 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Fixed

**Comment by ahamad-MS** [comment]
@limingu All comments are being addressed and all checks have passed.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py [hunk 3]
```diff
@@ -27,15 +29,33 @@
...
+
+def load_arguments(self, _):    # pylint: disable=too-many-statements
     with self.argument_context('eventgrid') as c:
         c.argument('resource_group_name', arg_type=resource_group_name_type)
         c.argument('location', arg_type=get_location_type(self.cli_ctx))
         c.argument('tags', arg_type=tags_type)
         c.argument('included_event_types', arg_type=included_event_types_type)
         c.argument('labels', arg_type=labels_type)
-        c.argument('endpoint_type', arg_type=get_enum_type(['webhook', 'eventhub', 'storagequeue', 'hybridconnection'], default='webhook'))
+        c.argument('endpoint_type', arg_type=get_enum_type(['webhook', 'eventhub', 'storagequeue', 'hybridconnection', 'servicebusqueue'], default='webhook'))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py]: Are these still not enums in your SDK??

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py [hunk 4]
```diff
@@ -6,22 +6,36 @@
...
     topic_type_mgmt_util = CliCommandType(
         operations_tmpl='azure.mgmt.eventgrid.operations#TopicTypesOperations.{}',
-        client_factory=topic_types_factory
+        client_factory=topic_types_factory,
+        client_arg_name='self'
     )
 
     with self.command_group('eventgrid topic', topics_mgmt_util, client_factory=topics_factory) as g:
-        g.show_command('show', 'get')
+        g.command('show', 'get')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Should be `show_command`

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py [hunk 5]
```diff
@@ -32,6 +46,24 @@ def load_command_table(self, _):
                                  setter_name='update',
                                  client_factory=topics_factory)
 
+    with self.command_group('eventgrid domain topic', domain_topics_mgmt_util, client_factory=domain_topics_factory) as g:
+        g.command('show', 'get')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Should be `show_command`

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py [hunk 6]
```diff
@@ -32,6 +46,24 @@ def load_command_table(self, _):
...
+        g.custom_command('list', 'cli_domain_topic_list')
+        g.custom_command('delete', 'cli_domain_topic_delete')
+        g.custom_command('create', 'cli_domain_topic_create_or_update')
+
+    with self.command_group('eventgrid domain', domains_mgmt_util, client_factory=domains_factory) as g:
+        g.command('show', 'get')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Should be `show_command`

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py [hunk 7]
```diff
@@ -50,15 +68,76 @@ def cli_topic_create_or_update(
...
+def cli_domain_topic_create_or_update(
+        client,
+        resource_group_name,
+        domain_name,
+        domain_topic_name):
+    async_domain_topic_create = client.create_or_update(
+        resource_group_name,
+        domain_name,
+        domain_topic_name)
+    return async_domain_topic_create.result()
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py]: The CLI handles this.

```Python
return client.create_or_update(…)
```
You should not be calling result yourself.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py [hunk 8]
```diff
@@ -50,15 +68,76 @@ def cli_topic_create_or_update(
...
+        location=location,
+        tags=tags)
 
-    created_topic = async_topic_create.result()
+    async_domain_create = client.create_or_update(
+        resource_group_name,
+        domain_name,
+        domain_info)
+    created_domain = async_domain_create.result()
+    return created_domain
```
Comment (tjprescott) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py]: Don't call result yourself. The CLI handles this. You simply return the value of the create_or_update call (which would be a poller).

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py (hunk 4 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Fixed

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py (hunk 5 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Fixed

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py (hunk 6 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/commands.py]: Fixed.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py (hunk 7 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py]: I was just following same call pattern in this file with existing code and same code is used in the extension. Anyhow, I am fixing this for the new code and all existing code. Kindly review these too.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py (hunk 8 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/custom.py]: Fixed here and for existing code.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py (hunk 3 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py]: yes. Same behavior as previous GA and extension.

**Comment by ahamad-MS** [comment]
@tjprescott All comments are addressed. Kindly review and merge as appropriate.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py [hunk 9]
```diff
@@ -44,31 +64,59 @@ def load_arguments(self, _):
...
+        c.argument('event_subscription_name', arg_type=name_type, help='Name of the event subscription.')
         c.argument('max_delivery_attempts', help="Maximum number of delivery attempts. Must be a number between 1 and 30.")
         c.argument('event_ttl', help="Event time to live (in minutes). Must be a number between 1 and 1440.")
         c.argument('deadletter_endpoint', help="The Azure resource ID of an Azure Storage blob container destination where EventGrid should deadletter undeliverable events for this event subscription.")
+        c.argument('advanced_filter', action=EventSubscriptionAddFilter, nargs='+')
+        c.argument('expiration_date', help="Date or datetime (in UTC, e.g. '2018-11-30T11:59:59+00:00' or '2018-11-30') after which the event subscription would expire. By default, there is no expiration for the event subscription.")
 
     with self.argument_context('eventgrid event-subscription create') as c:
-        c.argument('resource_group_name', deprecate_info=c.deprecate(redirect="--source-resource-id", expiration='2.1.0', hide=True), arg_type=resource_group_name_type)
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py]: Should the redirects be added back?

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py [hunk 10]
```diff
@@ -84,6 +199,8 @@
             For resource group:         --source-resource-id /subscriptions/{SubID}/resourceGroups/rg1
             For EventGrid topic:        --source-resource-id /subscriptions/{SubID}/resourceGroups/rg1/providers/Microsoft.EventGrid/topics/t1
             For storage account:        --source-resource-id /subscriptions/{SubID}/resourceGroups/rg1/providers/Microsoft.Storage/storageaccounts/sa1
+            For EventGrid domain:       --source-resource-id /subscriptions/{SubID}/resourceGroups/rg1/providers/Microsoft.EventGrid/domains/d1/topics/t1
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Minor typo: Since this is for a domain, the "/topics/t1" at the end should be removed.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py [hunk 11]
```diff
@@ -156,6 +297,8 @@
...
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Minor typo: Remove /topics/t1 suffix at the end.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py [hunk 12]
```diff
@@ -198,11 +341,21 @@
...
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Minor typo: Remove /topics/t1 suffix at the end.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py [hunk 13]
```diff
@@ -242,6 +395,8 @@
...
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Minor typo: Remove /topics/t1 suffix at the end.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py [hunk 14]
```diff
@@ -282,6 +449,8 @@
...
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Minor typo: Remove /topics/t1 suffix at the end.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py [hunk 15]
```diff
@@ -27,15 +29,33 @@
...
+    arg_type=name_type,
+    options_list=['--domain-name'],
+    completer=get_resource_name_completion_list('Microsoft.EventGrid/domains'))
 
-def load_arguments(self, _):
+domain_topic_name_type = CLIArgumentType(
+    help='Name of the domain topic.',
+    arg_type=name_type,
+    options_list=['--domain-topic-name'],
+    completer=get_resource_name_completion_list('Microsoft.EventGrid/domains/topic'))
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py]: Consider if we really need to have a completer for domain topics (since there could be 100,000 + of them). Not sure how this is implemented under the covers. Also, if we need a completer, I guess it needs to be topics instead of topic.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/advanced_filter.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+                           STRINGNOTIN + "," + STRINGBEGINSWITH + "," +
+                           STRINGCONTAINS + "," + STRINGENDSWITH + "," +
+                           NUMBERGREATERTHAN + "," + NUMBERGREATERTHANOREQUALS + "," +
+                           NUMBERLESSTHAN + "," + NUMBERLESSTHANOREQUALS + "," + BOOLEQUALS + ".")
+        if namespace.advanced_filter is None:
+            namespace.advanced_filter = []
+        namespace.advanced_filter.append(advanced_filter)
+
+
+def _validate_only_single_value_is_specified(operator_type, values):
```
Comment (kalyanaj) [inline src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/advanced_filter.py]: I think this limit has been removed now.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py (hunk 14 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Fixed

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py (hunk 13 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: thx. Fixed

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py (hunk 15 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py]: Agreed. Trying it in shell.azure.com for domains, it seems it waits for all results to be shown before displaying them to user. Showing that for 100K domain topic is challenging. Removing it due to perf concern in enumerating the domain topics in more than 100K exists

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py (hunk 12 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: Thx. Fixed.

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py (hunk 11 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_help.py]: thx. fixed

**Comment by ahamad-MS** [comment]
@tjprescott  You are right about the enum value in EventGrid Swagger. But your initial question was about the generated Python SDK which lacks this value and hence we don't use it here. We started a thread with Python SDK team who owns the  generation of Python SDK to determine if there is a bug in the autorest. Added you to the thread for your awareness but this should be addressed separately so we don't block this PR for the GA version.

I re-recorded the tests and they run fine locally.

**Comment by ahamad-MS** [comment]
@tjprescott Can you please help us with the test failures? The tests are re-recorded multiple times and they run fine in offline mode. Not sure what is causing such failure here event after full merge/sync with upstream/dev branch

File: src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py (hunk 3 above)
Comment (ahamad-MS) [reply src/command_modules/azure-cli-eventgrid/azure/cli/command_modules/eventgrid/_params.py]: You are right about the enum value in EventGrid Swagger. But your initial question was about the generated Python SDK which lacks this value and hence we don't use it here. We started a thread with Python SDK team who owns the generation of Python SDK to determine if there is a bug in the autorest. Added you to the thread for your awareness but this should be addressed separately so we don't block this PR for the GA version.

**Comment by ahamad-MS** [comment]
@tjprescott All comments are addressed and tests is passing along with all checks.
